consensus/core/src/core_thread.rs

```diff
@@ -0,0 +1,183 @@
+use std::sync::Arc;
+use std::{collections::HashSet, thread};
+use thiserror::Error;
+use tokio::sync::oneshot::error::RecvError;
+use tokio::sync::{mpsc, oneshot};
+use tracing::warn;
+
+#[allow(unused)]
+pub(crate) struct CoreThreadDispatcherHandle {
+    sender: mpsc::Sender<CoreThreadCommand>,
+    join_handle: thread::JoinHandle<()>,
+}
+
+impl CoreThreadDispatcherHandle {
+    #[allow(unused)]
+    pub fn stop(self) {
+        // drop the sender, that will force all the other weak senders to not able to upgrade.
+        drop(self.sender);
+        self.join_handle.join().ok();
+    }
+}
+
+#[allow(unused)]
+struct CoreThread {
+    core: Core,
+    receiver: mpsc::Receiver<CoreThreadCommand>,
+    context: Arc<Context>,
+}
+
+impl CoreThread {
+    pub fn run(mut self) {
+        tracing::debug!("Started core thread");
+
+        while let Some(command) = self.receiver.blocking_recv() {
+            let _scope = monitored_scope("CoreThread::loop");
+            self.context.metrics.node_metrics.core_lock_dequeued.inc();
+            match command {
+                CoreThreadCommand::AddBlocks(blocks, sender) => {
+                    let missing_blocks = self.core.add_blocks(blocks);
+                    sender.send(missing_blocks).ok();
+                }
+                CoreThreadCommand::ForceNewBlock(round, sender) => {
+                    self.core.force_new_block(round);
+                    sender.send(()).ok();
+                }
+                CoreThreadCommand::GetMissing(sender) => {
+                    // TODO: implement the logic to fetch the missing blocks.
+                    sender.send(vec![]).ok();
+                }
+            }
+        }
+    }
+}
+
+#[derive(Clone)]
+#[allow(dead_code)]
+pub(crate) struct CoreThreadDispatcher {
+    sender: mpsc::WeakSender<CoreThreadCommand>,
+    context: Arc<Context>,
+}
+
+enum CoreThreadCommand {
+    /// Add blocks to be processed and accepted
+    AddBlocks(Vec<Block>, oneshot::Sender<Vec<BlockRef>>),
+    /// Called when a leader timeout occurs and a block should be produced
+    ForceNewBlock(Round, oneshot::Sender<()>),
+    /// Request missing blocks that need to be synced.
+    GetMissing(oneshot::Sender<Vec<HashSet<BlockRef>>>),
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum CoreError {
+    #[error("Core thread shutdown: {0}")]
+    Shutdown(RecvError),
+}
+
+#[allow(unused)]
+impl CoreThreadDispatcher {
+    pub fn start(core: Core, context: Arc<Context>) -> (Self, CoreThreadDispatcherHandle) {
+        let (sender, receiver) = mpsc::channel(32);
+        let core_thread = CoreThread {
+            core,
+            receiver,
+            context: context.clone(),
+        };
+        let join_handle = thread::Builder::new()
+            .name("consensus-core".to_string())
+            .spawn(move || core_thread.run())
+            .unwrap();
+        // Explicitly using downgraded sender in order to allow sharing the CoreThreadDispatcher but
```

Comment:
Can you remind me what other component would need to use the core thread dispatcher? Is it the `BlockFetcher`?

Reply:
Right now is only the `net_sync`. But I see other cases where we'll probably need it. For example we are looking to extract the leader timeout logic out of `net_sync`, so that component will need access to the dispatcher